File: python/3. Longest Substring Without Repeating Characters/longest_substring.py
def length_of_longest_substring(s: str) -> int:
    window = {}
    start = temp_len = max_len = 0

    for i in range(len(s)):
        if (type(window.get(s[i], False)) != bool):
            for j in range(start, window[s[i]]): window.pop(s[j], None)
            
            temp_len += start - window[s[i]]
            start = window[s[i]] + 1
        else:
            temp_len += 1
            max_len = max(max_len, temp_len)
        
        window[s[i]] = i
    
    return max_len


if __name__ == "__main__":
    s = "hello world!"

    print(length_of_longest_substring(s))

File: c++/3. Longest Substring Without Repeating Characters/longest_substring.cpp
#include<bits/stdc++.h>

using namespace std;

int length_of_longest_substring(string s) {
    int max_len = 0, temp_len = 0, start = 0;
    unordered_map<char, int> window;

    for(int i = 0; i < s.size(); i++) {
        auto it = window.find(s[i]);

        if(it == window.end()) {
            temp_len++;
            max_len = max(max_len, temp_len);
            window.insert(make_pair(s[i], i));
        } else {
            for(int j = start; j< window[s[i]]; j++) window.erase(s[j]);

            temp_len = temp_len + start - window[s[i]];
            start = window[s[i]] + 1;
            window[s[i]] = i;
        }
    }

    return max_len;
}

int main() {
    string s = "hello world!";
    string b = "nbcnuweuponrhczmck";

    cout<<length_of_longest_substring(b);
    return 0;
}